uc-crux-llvm/src/UCCrux/LLVM/FullType/CrucibleType.hs

```diff
@@ -81,55 +99,155 @@ toCrucibleType proxy =
+ppTypeTranslationError =
+  \case
+    NoCFG (L.Symbol name) -> "Couldn't find definition of function " <> name
+    BadLift name -> "Couldn't lift argument/return types to MemTypes for " <> name
+    BadLiftArgs (L.Symbol name) ->
+      "Wrong number of arguments after lifting declaration of " <> name
+    FullTypeTranslation (L.Ident ident) ->
+      "Couldn't find or couldn't translate type: " <> ident
+
+translateModuleDefines ::
+  forall arch.
   ( ?lc :: TypeContext,
     ArchOk arch
   ) =>
-  proxy arch ->
-  Ctx.Assignment CrucibleTypes.TypeRepr crucibleTypes ->
-  Ctx.Assignment (Const MemType) crucibleTypes ->
-  Either L.Ident (SomeAssign arch crucibleTypes)
-assignmentToFullType proxy crucibleTypes memTypes =
+  L.Module ->
+  ModuleTranslation arch ->
+  Either TypeTranslationError (TranslatedTypes arch)
+translateModuleDefines llvmModule trans =
   case makeModuleTypes ?lc of
-    Some moduleTypes ->
+    Some initialModuleTypes ->
+      let Identity (maybeResult, modTypes) =
+            runStateT
+              ( runExceptT
+                  ( mapM
+                      (translateDecl . LLVMTrans.declareFromDefine)
+                      (L.modDefines llvmModule)
+                  )
+              )
+              initialModuleTypes
+       in TranslatedTypes modTypes . DeclTypes . Map.fromList <$> maybeResult
+  where
+    translateDecl ::
+      L.Declare ->
+      ExceptT
+        TypeTranslationError
+        (StateT (ModuleTypes m) Identity)
+        (L.Symbol, FunctionTypes m arch)
+    translateDecl decl =
       do
-        (Some fullTypes, moduleTypes') <-
-          runStateT
-            ( Ctx.generateSomeM
-                (Ctx.sizeInt (Ctx.size crucibleTypes))
-                ( \idx ->
-                    case Ctx.intIndex idx (Ctx.size crucibleTypes) of
-                      Nothing -> panic "Impossible" ["Mismatched context size"]
-                      Just (Some idx') ->
-                        do
-                          Some fullTypeRepr <-
-                            toFullTypeM (getConst (memTypes Ctx.! idx'))
-                          pure $ Some fullTypeRepr
-                )
+        liftedDecl <-
+          case LLVMTrans.liftDeclare decl of
+            Left err -> throwError (BadLift err)
+            Right d -> pure d
+        Crucible.AnyCFG cfg <-
+          case Map.lookup (L.decName decl) (LLVMTrans.cfgMap trans) of
+            Just (_, anyCfg) -> pure anyCfg
+            Nothing -> throwError (NoCFG (L.decName decl))
+        let crucibleTypes = Crucible.cfgArgTypes cfg
+        let memTypes = fdArgTypes liftedDecl
+        let isVarArgs = fdVarArgs liftedDecl
+        unless
+          ( let numCrucibleTypes = Ctx.sizeInt (Ctx.size crucibleTypes)
+             in length memTypes == numCrucibleTypes
+                  || ( isVarArgs
+                         && length memTypes + 1 == numCrucibleTypes
+                     )
+          )
```

Comment:
Could we name this condition? It would make it easier to visually see that this can lead to an error

Reply:
Yeah, will do!